plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,567 @@
+		retryTimes++;
+		if (retryTimes < 10)
+			goto Retry;
+		else
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_wrf(FuAlgoltekUsbDevice *self,
+		    GByteArray *firmware_data,
+		    FuProgress *progress,
+		    GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) transfer_parameter = g_byte_array_new();
+	g_autoptr(GByteArray) part_firmware = g_byte_array_new();
+	guint rest_transfer_data_size = firmware_data->len;
+	guint data_addr = 0;
+	guint start_index = 0;
+	guint max_packet_size = 64;
+	guint transfer_data_size = 0;
+	guint count_check = 0;
+	guint16 value;
+	guint16 index;
+
+	while (rest_transfer_data_size > 0) {
+		count_check++;
+		data_addr = start_index;
+		if (rest_transfer_data_size > max_packet_size)
+			transfer_data_size = max_packet_size;
+		else
+			transfer_data_size = rest_transfer_data_size;
+
+		fu_byte_array_set_size(transfer_parameter, 4, 0);
+
+		if (count_check == 256 / max_packet_size)
+			transfer_parameter->data[0] = 1;
+		else
+			transfer_parameter->data[0] = 0;
+
+		fu_memwrite_uint24(transfer_parameter->data + 1, data_addr, G_BIG_ENDIAN);
+
+		value = fu_memread_uint16(transfer_parameter->data, G_BIG_ENDIAN);
+		index = fu_memread_uint16(transfer_parameter->data + 2, G_BIG_ENDIAN);
+
+		fu_byte_array_set_size(part_firmware, transfer_data_size, 0);
+
+		for (guint i = 0; i < transfer_data_size; i++)
+			part_firmware->data[i] = firmware_data->data[i + start_index];
+
+		if (!g_usb_device_control_transfer(usb_device,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+						   G_USB_DEVICE_RECIPIENT_INTERFACE,
+						   ALGOLTEK_WRF,
+						   value, /* value */
+						   index, /* index */
+						   part_firmware->data,
+						   part_firmware->len,
+						   NULL,
+						   ALGOLTEK_DEVICE_USB_TIMEOUT,
+						   NULL,
+						   error)) {
+			g_prefix_error(error, "failed to contact device: ");
+			return FALSE;
+		}
+
+		start_index += transfer_data_size;
+		rest_transfer_data_size -= transfer_data_size;
+		if (count_check == (256 / max_packet_size) || rest_transfer_data_size == 0) {
+			count_check = 0;
+			if (!fu_algoltek_device_status_check(self, error))
+				return FALSE;
+		}
+		fu_progress_set_percentage_full(progress, start_index, firmware_data->len);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_setup(FuDevice *device, GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	g_autoptr(GByteArray) version_data = NULL;
+	g_autofree gchar *version_str = NULL;
+
+	/* UsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	version_data = algoltek_device_rdv(self, error);
+	version_str = g_strdup_printf("%s", version_data->data);
+
+	g_assert(self != NULL);
+	g_assert(usb_device != NULL);
```

Comment:
no g_assert in plugins please, it's not going to catch anything

Reply:
Already removed.